handler/src/main/java/io/netty/handler/ssl/SslHandler.java

```diff
@@ -1447,65 +1484,224 @@ private static ByteBuffer toByteBuffer(ByteBuf out, int index, int len) {
+                // we fire it through the pipeline as inbound error to be consistent with what we do in decode(...).
+                //
+                // This will also ensure we fail the handshake future and flush all produced data.
+                try {
+                    handleUnwrapThrowable(ctx, e);
+                } catch (Throwable cause) {
+                    safeExceptionCaught(cause);
                 }
+            } else {
+                setHandshakeFailure(ctx, e);
+                forceFlush(ctx);
+            }
+        }
 
-                tasks.add(task);
+        // Try to call exceptionCaught(...)
+        private void safeExceptionCaught(Throwable cause) {
+            try {
+                exceptionCaught(ctx, wrapIfNeeded(cause));
+            } catch (Throwable error) {
+                ctx.fireExceptionCaught(error);
             }
+        }
 
-            if (tasks.isEmpty()) {
-                return;
+        private Throwable wrapIfNeeded(Throwable cause) {
+            if (!inUnwrap) {
+                // If we are not in unwrap(...) we can just rethrow without wrapping at all.
+                return cause;
             }
+            // As the exception would have been triggered by an inbound operation we will need to wrap it in a
+            // DecoderException to mimic what a decoder would do when decode(...) throws.
+            return cause instanceof DecoderException ? cause : new DecoderException(cause);
+        }
+
+        private void tryDecodeAgain() {
+            try {
+                channelRead(ctx, Unpooled.EMPTY_BUFFER);
+            } catch (Throwable cause) {
+                safeExceptionCaught(cause);
+            } finally {
+                // As we called channelRead(...) we also need to call channelReadComplete(...) which
+                // will ensure we either call ctx.fireChannelReadComplete() or will trigger a ctx.read() if
+                // more data is needed.
+                channelReadComplete0(ctx);
+            }
+        }
 
-            final CountDownLatch latch = new CountDownLatch(1);
-            delegatedTaskExecutor.execute(new Runnable() {
-                @Override
-                public void run() {
-                    try {
-                        for (Runnable task: tasks) {
-                            task.run();
+        /**
+         * Executed after the wrapped {@code task} was executed via {@code delegatedTaskExecutor} to resume work
+         * on the {@link EventExecutor}.
+         */
+        private void resumeOnEventExecutor() {
+            assert ctx.executor().inEventLoop();
+
+            processTask = false;
+
+            try {
+                HandshakeStatus status = engine.getHandshakeStatus();
+                switch (status) {
+                    // There is another task that needs to be executed and offloaded to the delegatingTaskExecutor.
+                    case NEED_TASK:
```

Comment:
This should never happen, right? Maybe better to treat as some kind of state error?

Reply:
@njhill it could happen... there is nothing wrong with have another task to process after you run the current task.

Reply:
@normanmaurer I still don't quite see that.. won't it only reach here in the case all prior tasks have run followed by `getDelegatedTask` returning null and no other wrap/unwrap interactions having happened with the engine (since they happen on the event loop and guarded by `processTask` flag)?

Reply:
@njhill actually you are right I think it cant happen (at least with the current guard). That said I think there is nothing wrong with keep the code as it is as it not harms and if it ever changes all will "just work".